### ISSUE TITLE
ABIGen not generating function pointers for Go

### ISSUE THREAD

**Issue by jmc265** [ethereum/go-ethereum#17096]
#### System information

OS & Version: OSX
Commit hash : e916f9786dd318ec873cab21c8092d4da2c8dd54

#### Expected behaviour

ABIGen will generate a function pointer arg for the generated Go.

#### Actual behaviour

The generated Go file is not syntactically correct for the function argument.

#### Steps to reproduce the behaviour

1. Take this contract: https://ethfiddle.com/a7bXVPT940
2. Compile it `abigen -sol tmp.sol -pkg foo -out f.go`
3. Get errors of the form `./f.go:179:99: undefined: function`
4. Observe incorrect Go sytax in `f.go`:
```
func (_FunctionPointerTest *FunctionPointerTestTransactor) Test(opts *bind.TransactOpts, callback function) (*types.Transaction, error) {
        return _FunctionPointerTest.contract.Transact(opts, "test", callback)
}
```

**Comment by karalabe** [comment]
A `function` is not a documented ABI type https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI. I'm not sure when this got added into Solidity, or why it supports generating ABIs with it included. I don't think this is a feature that should be supported by the compiler.

**Comment by karalabe** [comment]
Ping @chriseth @axic

**Comment by chriseth** [comment]
I think this was added in 0.4.5 (2016-11-21). It is documented here: https://solidity.readthedocs.io/en/develop/abi-spec.html#types

Note that Solidity does not claim to implement **the** ABI, but it has some types that might be missing from other languages, so users of the abi json should be prepared to encounter unknown types from time to time. I know that this is not a good standardization mechanism, but I think EIPs were not yet in place at that time.

**Comment by chriseth** [comment]
Hm wait a second, the wiki page you link actually also lists `function`...

**Comment by Kegsay** [comment]
Presumably `abigen` can just replace the `function` type with `[24]byte` in Go? That would at least mean that projects which use this type can actually compile successfully.

**Comment by gballet** [comment]
Fixed by #19755
